server/src/main/java/org/opensearch/action/admin/cluster/remotestore/repository/RemoteStoreRepositoryRegistrationHelper.java

```diff
@@ -0,0 +1,364 @@
+                repository.getMetadata().name(),
+                repository.getMetadata().type()
+            );
+            repositoriesMetadata.add(newRepositoryMetadata);
+            repositories = new RepositoriesMetadata(repositoriesMetadata);
+        }
+        mdBuilder.putCustom(RepositoriesMetadata.TYPE, repositories);
+        return ClusterState.builder(currentState).metadata(mdBuilder).build();
+    }
+
+    private static ClusterState addRepository(
+        DiscoveryNode joiningNode,
+        ClusterState currentState,
+        RepositoriesService repositoriesService,
+        String repositoryNameAttributeKey
+    ) {
+        String repositoryName = joiningNode.getAttributes().get(repositoryNameAttributeKey);
+        ClusterState newState = updateClusterStateWithRepositoryMetadata(
+            currentState,
+            buildRepositoryMetadata(joiningNode, repositoryName),
+            repositoriesService
+        );
+
+        return newState;
+    }
+
+    private static ClusterState addRemoteStoreRepositories(
+        DiscoveryNode joiningNode,
+        ClusterState currentState,
+        RepositoriesService repositoriesService
+    ) {
+        ClusterState newState = ClusterState.builder(currentState).build();
+        List<DiscoveryNode> existingNodes = new ArrayList<>(currentState.nodes().getNodes().values());
+
+        if (joiningNode.equals(existingNodes.get(0)) || compareRemoteStoreNodeAttributes(joiningNode, existingNodes.get(0))) {
+            newState = addRepository(joiningNode, currentState, repositoriesService, REMOTE_STORE_SEGMENT_REPOSITORY_NAME_ATTRIBUTE_KEY);
+            newState = addRepository(
+                joiningNode,
+                ClusterState.builder(newState).build(),
+                repositoriesService,
+                REMOTE_STORE_TRANSLOG_REPOSITORY_NAME_ATTRIBUTE_KEY
+            );
+        }
+        return newState;
+    }
+
+    /**
+     * Validated or adds the remote store repositories to cluster state post its creation if the repositories doesn't
+     * exist. Returns the updated cluster state.
+     */
+    public static ClusterState validateOrAddRemoteStoreRepositories(
+        JoinTaskExecutor.Task joinTask,
+        ClusterState currentState,
+        RepositoriesService repositoriesService
+    ) {
+        ClusterState newState = ClusterState.builder(currentState).build();
+        /** Skipping the validation/addition of remote store repository in cluster state if task doesn't contain node
+         *  or node isn't a remote store node. **/
+        if (joinTask.node() != null && isRemoteStoreNode(joinTask.node())) {
+            if (JoinTaskExecutor.Task.ELECT_LEADER_TASK_REASON.equals(joinTask.reason())) {
+                newState = ClusterState.builder(addRemoteStoreRepositories(joinTask.node(), newState, repositoriesService)).build();
+            } else {
+                validateRemoteStoreRepositories(joinTask.node(), newState, repositoriesService);
```

Comment:
this sends out verify request to all nodes. For large clusters , this will kind of node join request . Even in case of transient node drops, this will increase the overall recovery times . Do we need to create a new verify method which just validates the newly joining node ?